lib/nss.c

```diff
@@ -131,6 +131,11 @@ void nss_init(const char *nsswitch_path) {
 		fprintf(shadow_logfd, "%s did not provide @find_subid_owners@\n", libname);
 		goto close_lib;
 	}
+	subid_nss->free= dlsym(h, "shadow_subid_free");
+	if (!subid_nss->free) {
```

Comment:
Isn't this conditional redundant with the one in `lib/subordinateio.c`?

Reply:
No, it is not. Consider the case when `libsubid_<plugin>.so` conforms to the current API and an application (e.g. `getsubids`) relies on this pull request. Then this conditional allows to deallocate memory with the new API while a plugin allocates a memory block in the assumption that the block will be `free`'d.

In other words, if a plugin uses the old API to allocate memory, it assumes that the memory will be `free`'d. At the same time, `getsubids` can use the new API to deallocate the memory since this conditional falls back to `free`. This ensures backward compatibility in this scenario.

On the other hand, we could make it plain and simple. If symbol `shadow_subid_free` is not exported then just return an error.

Reply:
Hmmm, I'm not fresh now and don't understand.  I'll let @hallyn review this, or maybe myself another day.  Thanks!  :)

Reply:
Let's raise an error when loading the plugin if it's not provided.

I was originally going to say fallback is fine, but it seems too likely that some provider will not know that they need to provide it...